source/opt/copy_prop_arrays.cpp

```diff
@@ -580,6 +591,39 @@ void CopyPropagateArrays::UpdateUses(Instruction* original_ptr_inst,
   for (auto pair : uses) {
     Instruction* use = pair.first;
     uint32_t index = pair.second;
+
+    if (use->GetOpenCL100DebugOpcode() != OpenCLDebugInfo100InstructionsMax) {
+      switch (use->GetOpenCL100DebugOpcode()) {
+        case OpenCLDebugInfo100DebugDeclare: {
+          context()->ForgetUses(use);
```

Comment:
Might be worth having a comment on what you are doing and why.  It might not be clear to other why you have to change the DebugDeclare to a DebugValue.

Reply:
I don't think the explanation you added makes sense.  Someone might ask why you cannot replace one pointer with another?  `new_ptr_inst` is not a value it is a pointer.  Please try again.

Reply:
After considering this more, I am confused what will be the correct approach to handle this case.
```
ImageArray input;
void main() {
  ImageArray foo;
  foo = input;
  // assume that value(foo) == value(input) == X
  ... use foo ...
  update value(input) = Y;
}

// Before copy-prop-array
%input = OpVariable %ptr_image_array_Input Input
...
%main = OpFunction
%foo = OpVariable %ptr_image_array_Function Function
OpStore %foo %input
DebugDeclare %dbg_foo %foo    --> value(foo) == value(input) == X here
... use %foo ...
... update %input to Y ...
// debugger_print(%dbg_foo) must be X here


// After copy-prop-array replaces %foo to %input
%input = OpVariable %ptr_image_array_Input Input
...
%main = OpFunction
DebugValue %dbg_foo %input Deref   // or DebugDeclare %dbg_foo %input
... use %foo ...
... update %input to Y ...
// debugger_print(%dbg_foo) must be Y, which is the value(input) not value(foo)
```

In the above example, since copy-prop-array pass replaces `%foo` to `%input`, the value inspection using a debugger at then end will be `Y` instead of `X`, where `Y` must be the value of `%input` and `X` must be the value of `%foo`.

In my opinion, if the `OpVariable` to be replaced has `DebugDeclare` or `DebugValue` with `Deref`, we have to skip replacing it with the other pointer because it will result in a wrong debugger value inspection result.

What do you think?